### PR TITLE
`#assert` parsing, typchecking, and SILGen (hidden under flag)

### PR DESCRIPTION
**Pull request by marcrasi**
Part 1 of mega-patch https://github.com/apple/swift/pull/19579.

One possible issue is that the flag doesn't disable the stuff in `utils/gyb_syntax_support/ExprNodes.py` and `utils/gyb_syntax_support/Token.py`, so if there's anything that picks up on that (e.g. code completion tools?), it won't respect the flag. Is this an issue? Anyone have suggestions on how to fix it?

### CODE REVIEW

File: lib/SILOptimizer/Mandatory/IRGenPrepare.cpp
Comment (harlanhaskins) [inline lib/SILOptimizer/Mandatory/IRGenPrepare.cpp]: `Builtin.poundAssert()`

Comment (marcrasi) [reply lib/SILOptimizer/Mandatory/IRGenPrepare.cpp]: yep, thanks, fixed!

**Review by ravikandhadai — APPROVED** [review]

**Comment by ravikandhadai** [comment]
LGTM

**Review by xedin — APPROVED** [review]
Sema related things look good to me!

File: lib/Parse/ParseExpr.cpp
Comment (rintaro) [inline lib/Parse/ParseExpr.cpp]: We usually use `parseMatchingToken()` for this.

File: lib/Parse/ParseExpr.cpp
Comment (rintaro) [inline lib/Parse/ParseExpr.cpp]: I don't think Parser should have the default message here. Could you make the default `StringRef()` and handle it in the actual assertion error logic?

File: lib/Parse/ParseExpr.cpp
Comment (rintaro) [inline lib/Parse/ParseExpr.cpp]: `#assert()` should really be modeled as a expression? From Sema code, I understand the type of `#assert(...)` expression is always `Void`. But I don't think `let foo: Void = #assert(true)` should even be accepted.
Why don't you model it as a statement?

File: lib/Parse/ParseExpr.cpp
Comment (rintaro) [inline lib/Parse/ParseExpr.cpp]: If `parseExpr()` has code completion status. We have to propagate it to the caller. Could you:
```cpp
if (conditionExprResult.hasCodeCompletion())
  return conditionExprResult;
```

File: lib/Parse/ParseExpr.cpp
Comment (rintaro) [inline lib/Parse/ParseExpr.cpp]: This is nice, but I don't think every expression parsing should have recovery logic on its own. We usually let `parseExprList()` or `parseBraceItems()` to recover from parse error.
Any particular reason to recover here? If there's no reason, diagnose and return `makeParserError()` should be sufficient.

**Comment by rintaro** [comment]
> One possible issue is that the flag doesn't disable the stuff in utils/gyb_syntax_support/ExprNodes.py and utils/gyb_syntax_support/Token.py, so if there's anything that picks up on that (e.g. code completion tools?), it won't respect the flag. Is this an issue? 

Currently there's no runtime/compile-time mechanism to enable/disable `libSyntax` node depending on flags. But I don't think that's a big issue. At least, we don't have such source tools which automatically lists up declared node kinds.

File: utils/gyb_syntax_support/ExprNodes.py
Comment (nkcsgexi) [inline utils/gyb_syntax_support/ExprNodes.py]: Could you add some #assert tests in round_trip_parse_gen.swift to ensure they are constructed as syntax nodes correctly?

File: lib/Sema/CSGen.cpp
Comment (nkcsgexi) [inline lib/Sema/CSGen.cpp]: Can we use `TypeChecker::typeCheckStmtCondition` to type check the pound condition just like for other statements if we module #assert as a statement instead of an expression like @rintaro suggested above? Generating constraint conditions seems to me an overkill for this case. cc: @xedin

Comment (xedin) [reply lib/Sema/CSGen.cpp]: I agree, it seems like we are to go statement route, it shouldn't require any type-checker changes, but should use existing `tyeCheckStmtCondition` mechanism.

File: lib/Parse/ParseExpr.cpp
Comment (marcrasi) [reply lib/Parse/ParseExpr.cpp]: I refactored it a bit and started using it.

File: lib/Parse/ParseExpr.cpp
Comment (marcrasi) [reply lib/Parse/ParseExpr.cpp]: done

**Review by marcrasi — COMMENTED** [review]
Thanks for all the comments! I addressed most of them, but I have 3 questions inline:
https://github.com/apple/swift/pull/19618#discussion_r228025801
https://github.com/apple/swift/pull/19618#discussion_r228025959
https://github.com/apple/swift/pull/19618#discussion_r228026185

File: utils/gyb_syntax_support/ExprNodes.py
Comment (nkcsgexi) [reply utils/gyb_syntax_support/ExprNodes.py]: Missing tags indicates unknown syntax kinds, which unfortunately means these syntax nodes are not constructed properly. You can use command `$./swift-macosx-x86_64/bin/swift-syntax-test -input-source-filename /tmp/t.swift -parse-gen -print-node-kind -print-trivial-node-kind` to print all unknown kinds and diagnose what's wrong with it.

File: lib/Parse/ParseExpr.cpp
Comment (rintaro) [reply lib/Parse/ParseExpr.cpp]: No. Returning error makes `parseBraceItems()` [automatically recover](https://github.com/apple/swift/blob/b01776164569cece2706d048707be37e970ce585/lib/Parse/ParseStmt.cpp#L468-L477). You don't need to touch `parseBraceItems()`.

File: lib/Parse/ParseStmt.cpp
Comment (rintaro) [inline lib/Parse/ParseStmt.cpp]: `return ParserStatus(conditionExprResult)` should be OK.

File: lib/Parse/ParseStmt.cpp
Comment (rintaro) [inline lib/Parse/ParseStmt.cpp]: Statement libSyntax parsing is a bit different from expression parsing. The caller (`parseStmt()`) have already [set up the context for the statement](https://github.com/apple/swift/blob/b01776164569cece2706d048707be37e970ce585/lib/Parse/ParseStmt.cpp#L547). So you should just set the syntax kind of it:
```cpp
SyntaxContext->setCreateSyntax(SyntaxKind::PoundAssertStmt);
```

File: utils/gyb_syntax_support/ExprNodes.py
Comment (marcrasi) [reply utils/gyb_syntax_support/ExprNodes.py]: Thanks, this helped me figure it out. The problem was that the test runs without "enable-experimental-pound-assert" so the parser bails out before parsing the whole statement, which confuses the syntax thing. I fixed it by making the parser parse the whole statement before bailing out.

File: lib/Parse/ParseStmt.cpp
Comment (marcrasi) [reply lib/Parse/ParseStmt.cpp]: thanks, done!

File: lib/Parse/ParseStmt.cpp
Comment (marcrasi) [reply lib/Parse/ParseStmt.cpp]: thanks, done!

**Comment by marcrasi** [comment]
Thanks for the answers! I've addressed the additional comments, and I don't have any questions left.

File: utils/gyb_syntax_support/StmtNodes.py
Comment (rintaro) [inline utils/gyb_syntax_support/StmtNodes.py]: `StringLiteralExpr ` -> `StringLiteralToken`.
This is why the last test case in `test/Syntax/round_trip_parse_gen.swift` isn't wrapped with `<PoundAssertStmt />`

**Comment by rintaro** [comment]
Parser changes LGTM otherwise!

File: utils/gyb_syntax_support/StmtNodes.py
Comment (marcrasi) [reply utils/gyb_syntax_support/StmtNodes.py]: oops, didn't notice that. thanks for catching. fixed!

**Comment by ravikandhadai** [comment]
@swift-ci please smoke test

**Comment by ravikandhadai** [comment]
@swift-ci Please smoke test

**Comment by marcrasi** [comment]
These failures look unrelated to what I did. I'll try merging with the latest master and running them again, to see what happens.

**Comment by marcrasi** [comment]
@swift-ci Please smoke test

**Comment by marcrasi** [comment]
@swift-ci Please smoke test

**Comment by marcrasi** [comment]
@swift-ci Please smoke test

**Comment by marcrasi** [comment]
@swift-ci Please smoke test

**Comment by ravikandhadai** [comment]
Yeah the failures seem related to the CI.

**Comment by marcrasi** [comment]
What's the procedure now? Do we need any more approvals? Should I clean up the commit history before we merge?

**Comment by devincoughlin** [comment]
Yes, it would be good to clean up the commit history before you merge and add an informative commit message for the entire commit.

File: include/swift/AST/DiagnosticsParse.def
Comment (marcrasi) [reply include/swift/AST/DiagnosticsParse.def]: done

File: lib/SILGen/SILGenStmt.cpp
Comment (marcrasi) [reply lib/SILGen/SILGenStmt.cpp]: Oh yep, I found some tests in our branch that I forgot to pull into this patch. I have pulled them in.

**Comment by marcrasi** [comment]
I addressed @devincoughlin's comments and cleaned up the commit history.

File: test/SILGen/pound_assert.swift
Comment (gottesmm) [inline test/SILGen/pound_assert.swift]: One suggestion. SIL emits the function name at the end brace of a sil function so you can pattern match the end of a function uniquely. Can you do that here? E.g.:

```
// CHECK: } // end sil function '$s12pound_assert13customMessage{{[_0-9a-zA-Z]*}}'
```

For both of the tests.

File: lib/SILGen/SILGenStmt.cpp
Comment (gottesmm) [inline lib/SILGen/SILGenStmt.cpp]: Can you change this to be a lower case s? (or even better stmt).

**Review by gottesmm — APPROVED** [review]
LGTM with my nits.

**Comment by ravikandhadai** [comment]
@marcrasi We have all the reviews. I think we are good to test and merge this, once the latest comments are fixed.

**Comment by marcrasi** [comment]
I addressed @gottesmm's comments and rebased to master.

I had to add a switch case to `lib/SILOptimizer/Transforms/AccessEnforcementReleaseSinking.cpp` for the rebase.

I'm not familiar with the process for running final tests and merging to master. @ravikandhadai, could you tell me how to do it, or just do it yourself? Thanks!

**Comment by ravikandhadai** [comment]
@swift-ci Please smoke test

**Comment by ravikandhadai** [comment]
@swift-ci Please smoke test

**Comment by ravikandhadai** [comment]
@marcrasi There is a new merge conflict, but it looks trivial to fix.

**Comment by marcrasi** [comment]
@swift-ci Please smoke test

**Comment by marcrasi** [comment]
rebased and fixed the conflict

**Comment by ravikandhadai** [comment]
@swift-ci please test benchmark

**Comment by ravikandhadai** [comment]
@swift-ci Please smoke benchmark

**Comment by ravikandhadai** [comment]
@swift-ci Please smoke benchmark

**Comment by ravikandhadai** [comment]
@swift-ci Please smoke test Linux platform

**Comment by ravikandhadai** [comment]
@swift-ci Please smoke test Linux platform

**Comment by ravikandhadai** [comment]
@swift-ci Please test and merge

**Comment by ravikandhadai** [comment]
@swift-ci Please test and merge

**Comment by ravikandhadai** [comment]
@swift-ci Please test Linux platform

**Comment by ravikandhadai** [comment]
@swift-ci Please test Linux Platform

**Comment by ravikandhadai** [comment]
@marcrasi Linux builds are failing repeatedly. They are running out of time. I am trying figure out what the issue could be.

**Comment by shahmishal** [comment]
@swift-ci test Linux

**Comment by marcrasi** [comment]
I see. Looks like it's timing out while running lldb tests? I'll see if I can reproduce on my linux machine.

**Comment by ravikandhadai** [comment]
@marcrasi It could be a problem/flakiness with the CI. @shahmishal has increased the timeout  and started the test again. Let us wait for this test.

**Comment by marcrasi** [comment]
Ok! Sounds good.

**Comment by ravikandhadai** [comment]
@swift-ci Please test and merge

**Pull request merged**
